pkg/operator/controller/gardenlet/add.go

```diff
@@ -0,0 +1,99 @@
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	seedmanagementv1alpha1 "github.com/gardener/gardener/pkg/apis/seedmanagement/v1alpha1"
+	predicateutils "github.com/gardener/gardener/pkg/controllerutils/predicate"
+	"github.com/gardener/gardener/pkg/utils/oci"
+)
+
+// ControllerName is the name of this controller.
+const ControllerName = "gardenlet"
+
+// AddToManager adds Reconciler to the given manager.
+func (r *Reconciler) AddToManager(ctx context.Context, mgr manager.Manager, virtualCluster cluster.Cluster) error {
+	if r.RuntimeCluster == nil {
+		r.RuntimeCluster = mgr
+	}
+	if r.RuntimeClient == nil {
+		r.RuntimeClient = mgr.GetClient()
+	}
+	if r.VirtualConfig == nil {
+		r.VirtualConfig = virtualCluster.GetConfig()
+	}
+	if r.VirtualAPIReader == nil {
+		r.VirtualAPIReader = virtualCluster.GetAPIReader()
+	}
+	if r.VirtualClient == nil {
+		r.VirtualClient = virtualCluster.GetClient()
+	}
+	if r.Clock == nil {
+		r.Clock = clock.RealClock{}
+	}
+	if r.GardenNamespaceTarget == "" {
+		r.GardenNamespaceTarget = v1beta1constants.GardenNamespace
+	}
+	if r.Recorder == nil {
+		r.Recorder = mgr.GetEventRecorderFor(ControllerName + "-controller")
+	}
+	if r.HelmRegistry == nil {
+		var err error
+		r.HelmRegistry, err = oci.NewHelmRegistry()
+		if err != nil {
+			return fmt.Errorf("failed creating new Helm registry: %w", err)
+		}
+	}
+
+	return builder.
+		ControllerManagedBy(mgr).
+		Named(ControllerName).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: pointer.IntDeref(r.Config.ConcurrentSyncs, 0),
+		}).
+		WatchesRawSource(
+			source.Kind(virtualCluster.GetCache(), &seedmanagementv1alpha1.Gardenlet{}),
+			&handler.EnqueueRequestForObject{},
+			builder.WithPredicates(
+				predicateutils.ForEventTypes(predicateutils.Create, predicateutils.Update),
+				&predicate.GenerationChangedPredicate{},
+				r.OperatorResponsiblePredicate(ctx),
+			),
+		).
+		Complete(r)
+}
+
+// OperatorResponsiblePredicate is a predicate for checking whether the Seed object has already been created for the
+// Gardenlet resource, and whether the kubeconfig secret ref has been removed.
+func (r *Reconciler) OperatorResponsiblePredicate(ctx context.Context) predicate.Predicate {
+	return predicate.NewPredicateFuncs(func(obj client.Object) bool {
+		gardenlet, ok := obj.(*seedmanagementv1alpha1.Gardenlet)
+		if !ok {
+			return false
+		}
+		return r.seedDoesNotExist(ctx, gardenlet) || gardenlet.Spec.KubeconfigSecretRef != nil
```

Comment:
I am surprised to see the check on the `KubeconfigSecretRef` here. It looks like we will not get into `r.reconcile(...)` due to the short-cut in `r.Reconcile(...)` (`if !r.seedDoesNotExist(ctx, gardenlet) {`).

How is this supposed to work?

Reply:
Nice catch :) I added this check in the `Reconcile` function only later and forgot about this predicate. Adapted now.